lib/content/write.js

```diff
@@ -1,160 +1,200 @@
+            cb()
+          },
+          (e) => {
+            ret.emit('error', e)
+          }
+        )
       })
-    })
-  })
-  ret.once('error', e => {
+    }
+  )
+  ret.once('error', (e) => {
     inputErr = e
   })
   return ret
 }
 
 function handleContent (inputStream, cache, opts, errCheck) {
-  return BB.using(makeTmp(cache, opts), tmp => {
+  return makeTmp(cache, opts).then((tmp) => {
     errCheck()
-    return pipeToTmp(
-      inputStream, cache, tmp.target, opts, errCheck
-    ).then(res => {
-      return moveToDestination(
-        tmp, cache, res.integrity, opts, errCheck
-      ).then(() => res)
-    })
+    return pipeToTmp(inputStream, cache, tmp.target, opts, errCheck)
+      .then((res) => {
+        return moveToDestination(
+          tmp,
+          cache,
+          res.integrity,
+          opts,
+          errCheck
+        ).then(() => res)
+      })
+      .then((result) => makeTmpDisposer(tmp, result))
+      .catch((err) => makeTmpDisposer(tmp, err, true))
   })
 }
 
 function pipeToTmp (inputStream, cache, tmpTarget, opts, errCheck) {
-  return BB.resolve().then(() => {
+  return Promise.resolve().then(() => {
     let integrity
     let size
-    const hashStream = ssri.integrityStream({
-      integrity: opts.integrity,
-      algorithms: opts.algorithms,
-      size: opts.size
-    }).on('integrity', s => {
-      integrity = s
-    }).on('size', s => {
-      size = s
-    })
+    const hashStream = ssri
+      .integrityStream({
+        integrity: opts.integrity,
+        algorithms: opts.algorithms,
+        size: opts.size
+      })
+      .on('integrity', (s) => {
+        integrity = s
+      })
+      .on('size', (s) => {
+        size = s
+      })
     const outStream = fs.createWriteStream(tmpTarget, {
       flags: 'wx'
     })
     errCheck()
-    return pipe(inputStream, hashStream, outStream).then(() => {
-      return { integrity, size }
-    }).catch(err => {
-      return rimraf(tmpTarget).then(() => { throw err })
-    })
+    return pipe(
+      inputStream,
+      hashStream,
+      outStream
+    )
+      .then(() => {
+        return { integrity, size }
+      })
+      .catch((err) => {
+        return rimraf(tmpTarget).then(() => {
+          throw err
+        })
+      })
   })
 }
 
 function makeTmp (cache, opts) {
   const tmpTarget = uniqueFilename(path.join(cache, 'tmp'), opts.tmpPrefix)
-  return fixOwner.mkdirfix(
-    cache, path.dirname(tmpTarget)
-  ).then(() => ({
+  return fixOwner.mkdirfix(cache, path.dirname(tmpTarget)).then(() => ({
     target: tmpTarget,
     moved: false
-  })).disposer(tmp => (!tmp.moved && rimraf(tmp.target)))
+  }))
+}
+
+function makeTmpDisposer (tmp, result, shouldThrow = false) {
+  const returnResult = () => {
+    if (shouldThrow) {
+      throw result
+    }
+    return result
+  }
+
+  if (tmp.moved) {
+    return returnResult()
+  }
+  return rimraf(tmp.target)
+    .then(() => returnResult())
+    .catch(() => returnResult())
```

Comment:
I would ideally say the same thing I mentioned previously in `lib/util/tmp.js` applies here but I see this one seems slightly more complex 😊